### PR TITLE
Improve reports performance

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Related to #5029 and #4782

I started playing with Aus production data locally, and had a look at reports again.

From my initial idea in Slack:

> I'm sure we're missing something. It's a problem of combining different sets, right? We build these sets like "editable_orders", "editable_line_items" and then combine them in a huge query but I think we can reduce the scope of those sets before building them...

> when we search reports we have filters that limit the results to a date range, set of distributors, etc.

> but it seems like we're applying that only after doing lots of big query building that includes things like "all the line items a user can edit, ever", which could be 50,000+

**TLDR;** if we add the scoping from the search filters into the logic of building sets like `editable_line_items` _before_ combining them into the query, it reduces page loads in various reports by **50-90%**, and the horrifying amount of `line_item` ids added to the `IN` clause is reduced to a reasonable amount.

#### What should we test?

- Reports are all working in general. 
- Should be loading significantly faster than before (on a server with lots of data).

This PR affects:
- **bulk_coop** reports
- **order_and_distributor** reports
- **orders_and_fulfillments** reports
- **packing** reports

#### Release notes

Improve reports performance

Changelog Category: Changed

### CODE REVIEW

File: app/services/permissions/order.rb [hunk 1]
```diff
@@ -2,24 +2,26 @@
 
 module Permissions
   class Order
-    def initialize(user)
+    def initialize(user, params = nil)
```
Comment (luisramos0) [inline app/services/permissions/order.rb]: we should only send params down to service classes if really needed. This is not the case. We dont need the full params object here, only params[:q]. I'd say you need to pass params[:q] and give it a good name at this service class level.
This params[:q] is specific to ransack, right? Maybe order_search_params or ransack_order_search_params?
This will make
def search_orders?
  order_search_params.present?
end

File: app/services/permissions/order.rb [hunk 2]
```diff
@@ -2,24 +2,26 @@
...
+      orders
     end
 
     # Any orders that the user can edit
     def editable_orders
-      Spree::Order.where(
-        managed_orders_where_values.
-          or(coordinated_orders_where_values)
-      )
+      orders = Spree::Order.where(managed_orders_where_values.or(coordinated_orders_where_values))
```
Comment (luisramos0) [inline app/services/permissions/order.rb]: I think this is so complex that it deserves the breaklines :-)

File: app/services/permissions/order.rb [hunk 3]
```diff
@@ -2,24 +2,26 @@
...
+      orders = orders.complete.not_state(:canceled).search(@params[:q]).result if search_orders?
```
Comment (luisramos0) [inline app/services/permissions/order.rb]: Is this change correct? Should the reports only list complete not cancelled orders?

Comment (luisramos0) [reply app/services/permissions/order.rb]: did you verify the SQL produced? Does it look good as part of the massive query?

File: lib/open_food_network/reports/line_items.rb [hunk 4]
```diff
@@ -12,9 +12,7 @@ def orders
       end
 
       def list(line_item_includes = nil)
-        line_items = @order_permissions.
-          visible_line_items.
-          merge(Spree::LineItem.where(order_id: orders.result))
+        line_items = @order_permissions.visible_line_items.in_orders(orders.result)
```
Comment (luisramos0) [inline lib/open_food_network/reports/line_items.rb]: I think this is not correct, I cant remember why. I think it's because of the SQL that visible_line_items generates, you cannot simply add a where clause, you need to use merge, which does something else, I cant remember what...

Comment (luisramos0) [reply lib/open_food_network/reports/line_items.rb]: I am sure I have tried this before and it didnt work. I cant remember why....

File: app/services/permissions/order.rb (hunk 3 above)
Comment (luisramos0) [reply app/services/permissions/order.rb]: if this is correct, I think it needs a spec.

File: lib/open_food_network/reports/line_items.rb (hunk 4 above)
Comment (Matt-Yorkley) [reply lib/open_food_network/reports/line_items.rb]: I know there can be issues with multiple #where calls directly in the same query chain, but I think its different if they're in named scopes?

File: app/services/permissions/order.rb (hunk 3 above)
Comment (Matt-Yorkley) [reply app/services/permissions/order.rb]: > did you verify the SQL produced? Does it look good as part of the massive query? 

I checked the actual output of the queries with real data, using rack-mini-profiler both before and after.

File: app/services/permissions/order.rb (hunk 1 above)
Comment (Matt-Yorkley) [reply app/services/permissions/order.rb]: Yes, I need to tidy up a bit.

File: lib/open_food_network/reports/line_items.rb (hunk 4 above)
Comment (luisramos0) [reply lib/open_food_network/reports/line_items.rb]: I didnt mean there would be a problem on the rails side building the query. I meant that I think the query that results from this is not correct for what we need.

File: app/services/permissions/order.rb (hunk 3 above)
Comment (luisramos0) [reply app/services/permissions/order.rb]: ok, great. that means you validated the replacement of the merge with the names scope as well.

Comment (Matt-Yorkley) [reply app/services/permissions/order.rb]: > Is this change correct? Should the reports only list complete not cancelled orders?

All the reports this PR affects use `Reports::LineItems` to get their line items, which uses this method (unchanged):
https://github.com/openfoodfoundation/openfoodnetwork/blob/981778bc98ce43dde6ff57da8272b997038745be/lib/open_food_network/reports/line_items.rb#L48-L50

File: lib/open_food_network/reports/line_items.rb (hunk 4 above)
Comment (Matt-Yorkley) [reply lib/open_food_network/reports/line_items.rb]: The #merge here gives us the intersection of both queries, right? So we could write it like this:
```ruby
line_items_in_orders_result = Spree::LineItem.where(order_id: orders.result)
visible_line_items = @order_permissions.visible_line_items
visible_line_items_in_orders_result = visible_line_items.merge(line_items_in_orders_result)
```

Is that not equivalent to `@order_permissions.visible_line_items.in_orders(orders.result)`?

**Comment by Matt-Yorkley** [comment]
Comparison of SQL output with this PR and without it (loading one day of data, non-superadmin user for a large enterprise, with AUS data):

https://gist.github.com/Matt-Yorkley/7a1a711178bc0e7b38e5a65f25953a61

**Comment by daniellemoorhead** [comment]
@mkllnk @sauloperez would be great to have another reviewer on this, feels super important to have this PR merged for the next release candidate.

File: app/services/permissions/order.rb (hunk 3 above)
Comment (Matt-Yorkley) [reply app/services/permissions/order.rb]: > if this is correct, I think it needs a spec

Specs added :+1:

File: lib/open_food_network/reports/line_items.rb (hunk 4 above)
Comment (Matt-Yorkley) [reply lib/open_food_network/reports/line_items.rb]: I called #to_sql on both versions and the output matches :+1:

File: app/services/permissions/order.rb [hunk 5]
```diff
@@ -35,6 +39,12 @@ def editable_line_items
 
     private
 
+    attr_reader :search_params
+
+    def search_orders?
+      search_params.present?
+    end
```
Comment (mkllnk) [inline app/services/permissions/order.rb]: I'm not sure with ransack but maybe this could be simpler. I'm imagining using this method in the cases above:

```suggestion
    def filtered_orders
      if search_params.present?
        Spree::Order.complete.not_state(:canceled).search(search_params).result
      else
        Spree::Order.scoped
      end
    end
```

```rb
    def editable_orders
      filtered_orders.where(
        managed_orders_where_values.
          or(coordinated_orders_where_values)
      )
    end
```

**Comment by Matt-Yorkley** [comment]
> When you measured the performance improvement, did you also check that the reports had exactly the same output?

Yes, I ran some reports with a lot of results, and diffed the final report output.

File: app/services/permissions/order.rb (hunk 5 above)
Comment (Matt-Yorkley) [reply app/services/permissions/order.rb]: I tried making these _exact_ changes and there were some issues with the specs; the returned results were not correct. 

I've added a new commit where it's refactored in a fairly similar way, adding a #filtered_orders method. Also renamed #search_orders? to #filter_orders?

**Comment by sauloperez** [comment]
@Matt-Yorkley would you mind running an `EXPLAIN ANALYZE` on the production server you took those queries you shared in the gist? It'd cool to then have the result on another gist plus a link to https://explain.dalibo.com/. This way we can use this PR as a baseline next time. I found that very useful while doing my investigations.

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley,

Amazing!

In AU-staging, as superadmin:
- bulk_coop reports and order_and_distributor reports ->> Load times dropped by a factor of 3. 
- orders_and_fulfillments reports and packing reports ->> still require more than 2 sec load time (snail). However, this was already the case before, and also I verified that in a regular admin account these reports work fine and are notriously faster.

Other reports seem to work fine as well.

I'd say this is a huge improvement :tada: 
Moving to ready-to-go.

**Comment by sauloperez** [comment]
I'm so looking forward to this!! :scream:
